### PR TITLE
perf(ext/crypto): use rayon threadpool for CPU bound tasks

### PR DESCRIPTION
**Pull request by littledivy**
Suggested by Tokio docs: https://docs.rs/tokio/latest/tokio/index.html#cpu-bound-tasks-and-blocking-code

> If your code is CPU-bound and you wish to limit the number of threads used to run it, you should use a separate thread pool dedicated to CPU bound tasks. For example, you could consider using the [rayon](https://docs.rs/rayon) library for CPU-bound tasks. It is also possible to create an extra Tokio runtime dedicated to CPU-bound tasks, but if you do this, you should be careful that the extra runtime runs only CPU-bound tasks, as IO-bound tasks on that runtime will behave poorly.

~2x improvement hashing sha256:

```
# main
deno run -A --unstable cli/bench/sha256.js 

time 109 ms rate 91743
time 110 ms rate 90909
time 114 ms rate 87719
time 113 ms rate 88495
time 113 ms rate 88495
```

```
# node
node cli/bench/sha256.js

time 74 ms rate 135135
time 59 ms rate 169491
time 59 ms rate 169491
time 59 ms rate 169491
time 58 ms rate 172413
```

```
# this patch
target/release/deno run -A --unstable cli/bench/sha256.js

time 70 ms rate 142857
time 60 ms rate 166666
time 59 ms rate 169491
time 59 ms rate 169491
time 57 ms rate 175438
```

On M1 Mac, memory usage stays the same around 20mb. CPU utilization is way better.

### CODE REVIEW

File: cli/bench/sha256.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  const start = Date.now();
+  for (let i = 0; i < count; i++) await fun();
+  const elapsed = Date.now() - start;
+  const rate = Math.floor(count / (elapsed / 1000));
+  console.log(`time ${elapsed} ms rate ${rate}`);
+  if (--total) queueMicrotask(() => bench(fun));
+}
+
+const c = typeof crypto !== "undefined" ? crypto : require("crypto").webcrypto;
+const x = new Uint8Array(1000);
```
Comment (billywhizz) [inline cli/bench/sha256.js:20]: should we try this with an array of random bytes? and different sizes? i can have a look over weekend if you like.

**Comment by dsherret** [comment]
Would be nice if we did this for formatting and linting too. Maybe I will do this after this is merged. I previously tried to introduce rayon there referencing the docs you mentioned, but this tokio_rayon crate didn't exist at the time (#7274).

**Comment by billywhizz** [comment]
this makes sense to me. this article that divy posted does a good job describing the approach. https://ryhl.io/blog/async-what-is-blocking/

so, the rule seems to be to use rayon for anything cpu bound and spawn blocking only for operations that might be blocked awaiting io or some system event, but not utilizing the cpu.

**Comment by bnoordhuis** [comment]
I'm trying to understand _why_ it's faster. Is computing the SHA-256 of 1,000 bytes a meaningful amount of CPU-intensive work or are you merely timing thread pool round-trip times?

tokio-rayon is really just rayon that sends the result back over a oneshot channel (i.e., very thin shim) so I can think of two reasons for the stark difference:

1. tokio has many  threads per core, rayon is thread per core, resulting in fewer threads fighting for CPU time. That should get reflected in the number of (in)voluntary context switches in `/usr/bin/time -v deno run bench.js`.

2. tokio's job queue is FIFO, rayon's is LIFO. Does `spawn_fifo()` make a difference? LIFO can, theoretically, result in starvation.

**Comment by billywhizz** [comment]
@bnoordhuis yes, my understanding was that rayon will create one thread per core and pin those threads so any work on them should have less context switches. it may just be in this benchmark there is no other work happening and there is affinity between the main thread and the thread doing the async work? this makes a big difference for these kinds of micro benchmarks, but might not have much impact in real world on a system under load?

**Comment by littledivy** [comment]
Closing because stale but we should revisit this later.

**Pull request closed without merging**